src/lib/child_processes/child_processes.ml

```diff
@@ -0,0 +1,444 @@
+(** Management of starting, tracking, and killing child processes. *)
+
+open Core
+open Async
+open Pipe_lib
+module Termination = Termination
+
+exception Child_died
+
+type t =
+  { process: Process.t
+  ; stdout_pipe: string Strict_pipe.Reader.t
+  ; stderr_pipe: string Strict_pipe.Reader.t
+  ; stdin: Writer.t
+  ; terminated_ivar: Unix.Exit_or_signal.t Ivar.t
+  ; mutable killing: bool
+  ; mutable termination_response:
+      [ `Always_raise
+      | `Raise_on_failure
+      | `Handler of killed:bool -> Unix.Exit_or_signal.t -> unit Deferred.t
+      | `Ignore ] }
+
+let stdout_lines : t -> string Strict_pipe.Reader.t = fun t -> t.stdout_pipe
+
+let stderr_lines : t -> string Strict_pipe.Reader.t = fun t -> t.stderr_pipe
+
+let stdin : t -> Writer.t = fun t -> t.stdin
+
+let termination_status : t -> Unix.Exit_or_signal.t option =
+ fun t -> Ivar.peek t.terminated_ivar
+
+(** Try running [f] until it returns [Ok], returning the first [Ok] or [Error]
+    if all attempts fail. *)
+let keep_trying :
+    f:('a -> 'b Deferred.Or_error.t) -> 'a list -> 'b Deferred.Or_error.t =
+ fun ~f xs ->
+  let open Deferred.Let_syntax in
+  let rec go e xs : 'b Deferred.Or_error.t =
+    match xs with
+    | [] ->
+        return e
+    | x :: xs -> (
+        match%bind f x with
+        | Ok r ->
+            return (Ok r)
+        | Error e ->
+            go (Error e) xs )
+  in
+  go (Or_error.error_string "empty input") xs
+
+(* Unfortunately, `dune runtest` runs in a pwd deep inside the build directory.
+   This hack finds the project root by recursively looking for the dune-project
+   file. *)
+let get_project_root () =
```

Comment:
`get_git_root`

You don't want to run `git` for this: `git rev-parse --show-toplevel`

Reply:
You could be running the tests from an sdist